### PR TITLE
[podmanreceiver] Add metrics and resource metadata

### PR DESCRIPTION
**Pull request by rogercoll**
**Description:** <Describe what has changed.>
- Adds "metadata.yml" file to autogenerate metrics and resources.
- [Update: not done in this PR] Fixes invalid network metrics: "rx -> input" and "tx -> output"

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/28640

**Testing:** Previous tests preserved.

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

File: receiver/podmanreceiver/metadata.yaml
Comment (fatsheep9146) [inline receiver/podmanreceiver/metadata.yaml:108]: It seems that the format of link is not valid. Should it be like this? [More docs](https://www.kernel.org/doc/Documentation/cgroup-v1/blkio-controller.txt).

**Comment by fatsheep9146** [comment]
Could you also update the `metric` section in ReadMe.md to let user to see the `documentation.md` to check about the metrics supported by podman receiver?

File: receiver/podmanreceiver/metadata.yaml
Comment (fatsheep9146) [inline receiver/podmanreceiver/metadata.yaml:72]: I think the memory related metrics' unit should be `By`.

Comment (fatsheep9146) [reply receiver/podmanreceiver/metadata.yaml:72]: You can take https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/hostmetricsreceiver/internal/scraper/memoryscraper/metadata.yaml as example

File: receiver/podmanreceiver/metadata.yaml
Comment (rogercoll) [reply receiver/podmanreceiver/metadata.yaml]: Fixed in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30232/commits/df281be2f8410406c032f3c55dfa8e39a931cd44

**Comment by rogercoll** [comment]
> Could you also update the `metric` section in ReadMe.md to let user to see the `documentation.md` to check about the metrics supported by podman receiver?

Definitely, thanks for the feedback. Added in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30232/commits/df281be2f8410406c032f3c55dfa8e39a931cd44

File: receiver/podmanreceiver/metadata.yaml
Comment (rogercoll) [reply receiver/podmanreceiver/metadata.yaml]: Makes sense, fixed in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30232/commits/df281be2f8410406c032f3c55dfa8e39a931cd44

File: receiver/podmanreceiver/metadata.yaml
Comment (fatsheep9146) [inline receiver/podmanreceiver/metadata.yaml:121]: This line should also be fixed

Comment (rogercoll) [reply receiver/podmanreceiver/metadata.yaml]: Fixed in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30232/commits/5f3528cebd4e3ec56b485b9f989643c921319754. Thanks

File: receiver/podmanreceiver/metadata.yaml
Comment (fatsheep9146) [inline receiver/podmanreceiver/metadata.yaml:39]: https://opentelemetry.io/docs/specs/semconv/system/system-metrics/#metric-systemcputime
I'm not sure if cputime unit should be `ns` or `s`, @open-telemetry/collector-approvers should this metric use `s` as unit?

File: cmd/mdatagen/validate_test.go
Comment (mackjmr) [inline cmd/mdatagen/validate_test.go:111]:
```suggestion
		"container.cpu.utilization":           {"docker_stats", "kubeletstats"},
```

Comment (rogercoll) [reply cmd/mdatagen/validate_test.go:111]: Thanks!

File: receiver/podmanreceiver/metadata.yaml
Comment (rogercoll) [reply receiver/podmanreceiver/metadata.yaml:39]: I think that cputime metric's unit is in seconds, as this is how many OSes report it (/proc/stats). But containers are handled by cgroup controllers, which allow for [better precision](https://www.kernel.org/doc/Documentation/cgroup-v2.txt). I would say not to lose precision, instead add container metrics into the semantic convention.

Comment (fatsheep9146) [reply receiver/podmanreceiver/metadata.yaml:39]: Your comments also make sense, I will pull this discussion to slacks, to get more input. @rogercoll

Comment (mackjmr) [reply receiver/podmanreceiver/metadata.yaml:39]: The docker stats receiver collects `container.cpu.usage.system` as well: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/dockerstatsreceiver/metadata.yaml#L63-L67, and the unit is `ns`. IMO we should also use `ns` here to stay consistent between the two.

Comment (rogercoll) [reply receiver/podmanreceiver/metadata.yaml:39]: Sounds good! Does it make sense to start working on the container's semantic convention PR? (feel free to add me into the slack thread _@rogercoll_)

Comment (fatsheep9146) [reply receiver/podmanreceiver/metadata.yaml:39]: No one responses for that for a day. And as @mackjmr  said docker stats receiver also take `ns` as unit, so I think this is ok to approve.

**Review by fatsheep9146 — APPROVED** [review]

**Review by mackjmr — APPROVED** [review]

**Comment by rogercoll** [comment]
@fatsheep9146 Is there anything missing for the merge?

**Comment by fatsheep9146** [comment]
> @fatsheep9146 Is there anything missing for the merge?

Sorry for delay, I got advise from maintainers @dmitryax about the unit, he suggested that we conform to the semantic conventions.

image

I suggested we you `s` instead of `ns` first and open a pr in semantic convention related repo to try to prove the `ns` is better.
@rogercoll

**Comment by rogercoll** [comment]
@fatsheep9146 Cpu metrics precision changed in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30232/commits/a5a00992ee43a646b6873608dbcfe31da59fafcb

**Comment by fatsheep9146** [comment]
@dmitryax could you help finally review about this pr?

File: .chloggen/add_podman_metadata.yaml
Comment (dmitryax) [inline .chloggen/add_podman_metadata.yaml:10]: It's unclear from the description what metrics and metadata are added. Can we make the migration to metadata.yaml as a no-op change for the end user and handle any other changes separately?

Comment (rogercoll) [reply .chloggen/add_podman_metadata.yaml]: Make sense to me, let's make the "networking" changes into another PR. Removed in: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30232/commits/91e24ff4bbbf8d199ab88b943c887781d27672c1

File: cmd/mdatagen/validate_test.go
Comment (mx-psi) [inline cmd/mdatagen/validate_test.go:114]: Is it okay to report the same metric with different units?

As discussed above, on the Docker stats receiver we have nanoseconds https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/facd3696decbc50cc906645cbd2a0d47f515c18c/receiver/dockerstatsreceiver/documentation.md?plain=1#L41-L47
while here we use seconds https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/91e24ff4bbbf8d199ab88b943c887781d27672c1/receiver/podmanreceiver/documentation.md?plain=1#L65-L71

Is that okay? If seconds is the right unit, shouldn't we use it on the Docker stats receiver as well?

Comment (rogercoll) [reply cmd/mdatagen/validate_test.go:114]: I don't think so, I would not use second's precision for the sake of convenience at the expense of precision. Instead, we should endeavor to establish distinct conventions specifically tailored to containers. Should we wait for the container's semantic convention https://github.com/open-telemetry/semantic-conventions/pull/282 (nanoseconds)?

Comment (mx-psi) [reply cmd/mdatagen/validate_test.go:114]: > I don't think so, I would not use second's precision for the sake of convenience at the expense of precision. Instead, we should endeavor to establish distinct conventions specifically tailored to containers.

My objection here is with the use of different units on each metric, I would expect them to have the same unit (whether it is nanoseconds or seconds I agree is something we can leave https://github.com/open-telemetry/semantic-conventions/pull/282 to decide on)

**Comment by MovieStoreGuy** [comment]
Can I ask that the files in conflict get fixed up, outside of that, I don't see any outstanding issues that would block this PR?

**Comment by fatsheep9146** [comment]
> Can I ask that the files in conflict get fixed up, outside of that, I don't see any outstanding issues that would block this PR?

I think we still wait for the https://github.com/open-telemetry/semantic-conventions/pull/282 to be merged.

@MovieStoreGuy

**Comment by fatsheep9146** [comment]
just ping jsuereth on https://github.com/open-telemetry/semantic-conventions/pull/282 to merge the pr.

**Comment by fatsheep9146** [comment]
@rogercoll could you push forward this pr since https://github.com/open-telemetry/semantic-conventions/pull/282 is merge?

**Comment by rogercoll** [comment]
After the merge of https://github.com/open-telemetry/semantic-conventions/pull/282, some additional changes will still be required:

- [ ] **IO**: From `container.blockio.io_service_bytes_recursive` to `container.disk.io` 
- [ ] **Network**: From `container.network.io.usage` to `container.network.io`
- [ ] **Memory**:
  - [ ] From `container.memory.usage.total` to `container.memory.usage`
  - [ ] Add `container.memory.limit` to semantic conventions
  - [ ] Add `container.memory.utilization` to semantic conventions
- [ ] **CPU**:
  - [ ] From `container.cpu.usage.system` and `container.cpu.usage.total` to `container.cpu.time`
  - [ ] Add `container.cpu.logical_number` attribute to semantic conventions.
  - [ ] Add `container.cpu.utilization` to semantic conventions.

**Comment by MovieStoreGuy** [comment]
My preference would be to split the changes to keep change sizes small, are you okay with that @fatsheep9146 ?

**Comment by fatsheep9146** [comment]
> My preference would be to split the changes to keep change sizes small, are you okay with that @fatsheep9146 ?

I agreed. @rogercoll could you fix the conflicts, so I could push this pr to be merged ASAP.

**Comment by rogercoll** [comment]
> I agreed. @rogercoll could you fix the conflicts, so I could push this pr to be merged ASAP.

Done, I also updated the changelog file to a "breaking" change type due to the cpu precision fix (ns -> s).

**Comment by rogercoll** [comment]
If we prefer a totally no-op change PR, I could rollback the https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30232/commits/a5a00992ee43a646b6873608dbcfe31da59fafcb commit and push it into a different PR. Let me know what you think

**Comment by angristan** [comment]
👋 Sorry to dig up this PR! Is there a reason we went with `s` as the unit for the CPU time metrics?

I see this was discussed in these two comments on the PR:

- https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30232#discussion_r1469848897
- https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30232#discussion_r1439147724

Right now, for the same metric names, the `podmanreceiver` uses `s` and `dockerstatsreceiver` uses `ns` and the [semconv](https://opentelemetry.io/docs/specs/semconv/system/container-metrics/#metric-containercpuusage) suggests something else entirely 🤔

**Comment by rogercoll** [comment]
> 👋 Sorry to dig up this PR! Is there a reason we went with `s` as the unit for the CPU time metrics?
> 
> I see this was discussed in these two comments on the PR:
> 
>     * [[podmanreceiver] Add metrics and resource metadata #30232 (comment)](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30232#discussion_r1469848897)
> 
>     * [[podmanreceiver] Add metrics and resource metadata #30232 (comment)](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30232#discussion_r1439147724)
> 
> 
> Right now, for the same metric names, the `podmanreceiver` uses `s` and `dockerstatsreceiver` uses `ns` and the [semconv](https://opentelemetry.io/docs/specs/semconv/system/container-metrics/#metric-containercpuusage) suggests something else entirely 🤔

That's right, there's a misalignment between the two receivers and the semantic conventions. The fix is being discussed in this issue https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/31649

**Pull request merged**
